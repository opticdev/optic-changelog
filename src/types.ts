export type Changelog = {
  data: {
    endpoints: {
      change: {
        category: string
      }
      path: string
      method: string
    }[]
  }
}

export type JobInputs = {
  repoToken?: string
  subscribers: string[]
  opticSpecPath: string
}

export type RepoInfo = {
  owner: string
  repo: string
  prNumber?: number
  headSha: string
}

export type PrComment = {
  id: number
  body: string
}

export type PrInfo = {
  baseSha: string
  baseBranch: string
}

export declare interface IGitProvider {
  getFileContent(sha: string, path: string): Promise<string>
  getPrInfo(prNumber: number): Promise<PrInfo>
  updatePrComment(commentId: number, body: string): Promise<void>
  createPrComment(prNumber: number, body: string): Promise<void>
  getPrBotComments(prNumber: number): Promise<PrComment[]>
}

export declare interface IJobRunner {
  debug(message: string): void
  info(message: string): void
  setFailed(message: string | Error): void
}
