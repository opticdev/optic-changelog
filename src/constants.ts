export const API_BASE = process.env.OPTIC_API_BASE || 'https://api.useoptic.com'
export const CLOUD_SPEC_VIEWER_BASE =
  process.env.CLOUD_SPEC_VIEWER_BASE || 'https://spec.useoptic.com'
export const COMMENT_HEADER_IMG =
  'https://raw.github.com/opticdev/optic/develop/assets/changelog_header.png'
export const SENTRY_DSN =
  'https://b5d9462fe8e741c1a01c1f0ec8562241@o446328.ingest.sentry.io/5741686'
