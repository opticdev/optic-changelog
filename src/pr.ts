import {Changelog} from './types'

const REGEX = /\n\n<!-- optic = (.*) -->/

export function isOpticComment(body: string): boolean {
  return body.match(REGEX) ? true : false
}

export function getMetadata(body: string): any {
  const match = body.match(REGEX)

  if (match) {
    return JSON.parse(match[1])
  }

  return {}
}

export function setMetadata(body: string, data: any): string {
  let currentData = {}

  const bodyText = body.replace(REGEX, (_, json) => {
    currentData = JSON.parse(json)
    return ''
  })

  return `${bodyText}\n\n<!-- optic = ${JSON.stringify({
    ...currentData,
    ...data
  })} -->`
}

export function generateCommentBody(
  changes: Changelog,
  subscribers: string[]
): string {
  const results = {
    added: 0,
    updated: 0,
    removed: 0
  }

  for (const endpoint of changes.data.endpoints) {
    switch (endpoint.change.category) {
      case 'added':
        results.added++
        break
      case 'updated':
        results.updated++
        break
      case 'removed':
        results.removed++
        break
    }
  }

  const timestamp = new Date()
    .toISOString()
    .replace(/T/, ' ')
    .replace(/\..+/, '')

  const baseBody = `## Optic Changelog
  
* Endpoints added: ${results.added}
* Endpoints updated: ${results.updated}

Last updated: ${timestamp}

[View documentation](${changes.data.opticUrl})
`

  if (subscribers.length) {
    const subscriberText = subscribers
      .map(subscriber => `@${subscriber}`)
      .join(', ')
    return `${baseBody}
---

Pinging subscribers ${subscriberText}`
  }

  return baseBody
}
