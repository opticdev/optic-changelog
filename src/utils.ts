import * as Sentry from '@sentry/node'
import {SpanContext} from '@sentry/types/dist'
import {Span} from '@sentry/tracing'

export async function sentryInstrument<T>(
  spanParams: Partial<SpanContext>,
  cb: (
    transaction: ReturnType<typeof Sentry['startTransaction']>,
    span: Span
  ) => Promise<T>
): Promise<T> {
  const existingTxn = !!Sentry.getCurrentHub().getScope()?.getTransaction()
  const transaction =
    Sentry.getCurrentHub().getScope()?.getTransaction() ??
    Sentry.startTransaction({
      op: 'gitbot_run',
      name: 'Gitbot Run'
    })

  const existingSpan = Sentry.getCurrentHub().getScope()?.getSpan()

  const span = existingSpan
    ? existingSpan.startChild(spanParams)
    : transaction.startChild(spanParams)

  Sentry.getCurrentHub().configureScope(s => s.setSpan(span))
  try {
    return await cb(transaction, span)
  } catch (e) {
    Sentry.captureException(e)
    throw e
  } finally {
    span.finish()
    if (existingSpan) {
      Sentry.getCurrentHub().configureScope(s => s.setSpan(existingSpan))
    } else {
      Sentry.getCurrentHub().configureScope(s => s.setSpan(transaction))
    }
    if (!existingTxn) {
      // We started the txn, so we need to finish it
      transaction.finish()
    }
  }
}
