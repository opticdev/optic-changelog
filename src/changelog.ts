import {Changelog} from './types'
import {setMetadata, isOpticComment, generateCommentBody} from './pr'

export async function runOpticChangelog({
  subscribers,
  opticSpecPath,
  gitProvider,
  headSha,
  baseSha,
  baseBranch,
  prNumber,
  jobRunner
}): Promise<void> {
  let headContent: string, baseContent: string

  // This may fail if the file was moved or it's a new Optic setup
  try {
    headContent = await gitProvider.getFileContent(headSha, opticSpecPath)
  } catch (error) {
    // Failing silently here
    jobRunner.info(
      `Could not find the Optic spec in the current branch. Looking in ${opticSpecPath}.`
    )
    return
  }

  // This may fail if the file was moved or it's a new Optic setup
  try {
    baseContent = await gitProvider.getFileContent(baseSha, opticSpecPath)
  } catch (error) {
    // Failing silently here
    jobRunner.info(
      `Could not find the Optic spec in the base branch ${baseBranch}. Looking in ${opticSpecPath}.`
    )
    return
  }

  // TODO: use new changelog library here
  const changes = getChangelogData({
    from: JSON.parse(baseContent),
    to: JSON.parse(headContent)
  })

  if (changes.data.endpoints.length === 0) {
    jobRunner.info('No API changes in this PR.')
    return
  }

  const message = generateCommentBody(changes, subscribers)
  const body = setMetadata(message, {})

  jobRunner.debug('Created body for comment')
  jobRunner.debug(body)

  try {
    // TODO: probably should be simplified a bit
    const existingBotComments = (
      await gitProvider.getPrBotComments(prNumber)
    ).filter(comment => isOpticComment(comment.body!))

    if (existingBotComments.length) {
      const comment = existingBotComments[0]
      jobRunner.debug(`Updating comment ${comment.id}`)
      // TODO: need to pull out metadata and combine with new (maybe)
      await gitProvider.updatePrComment(comment.id, body)
    } else {
      jobRunner.debug(`Creating comment for PR ${prNumber}`)
      await gitProvider.createPrComment(prNumber, body)
    }
  } catch (error) {
    jobRunner.info(
      `There was an error creating or updating a PR comment. Error message: ${error.message}`
    )
    return
  }
}

// TODO: this is fake data for now
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function getChangelogData(options: object): Changelog {
  return {
    data: {
      opticUrl: 'https://example.com',
      endpoints: [
        {
          change: {
            category: 'added'
          },
          path: '/foo',
          method: 'get'
        },
        {
          change: {
            category: 'updated'
          },
          path: '/bar',
          method: 'post'
        }
      ]
    }
  }
}
