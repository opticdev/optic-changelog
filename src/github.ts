import * as core from '@actions/core'
import * as github from '@actions/github'
import {GitHub} from '@actions/github/lib/utils'
import {IGitProvider, JobInputs, RepoInfo, PrInfo, PrComment} from './types'

export function getJobInputs(): JobInputs {
  const repoToken = core.getInput('GITHUB_TOKEN') || process.env['GITHUB_TOKEN']

  const subscribers = core
    .getInput('subscribers')
    .split(',')
    .map(subscriber => subscriber.trim())

  const opticSpecPath = core.getInput('OPTIC_SPEC_PATH')
  const opticApiKey = core.getInput('OPTIC_API_KEY')

  return {repoToken, subscribers, opticSpecPath, opticApiKey}
}

export function getRepoInfo(): RepoInfo {
  const {
    payload: {repository, pull_request: pullRequest},
    sha: headSha
  } = github.context
  const {full_name: repoFullName = ''} = repository ?? {}
  const [owner, repo] = repoFullName.split('/')

  return {
    prNumber: pullRequest?.number,
    owner,
    repo,
    headSha
  }
}

export class GitHubRepository implements IGitProvider {
  constructor(
    private octokit: InstanceType<typeof GitHub>,
    public owner: string,
    public repo: string
  ) {}

  async getFileContent(sha: string, path: string): Promise<string> {
    const response = await this.octokit.repos.getContent({
      owner: this.owner,
      repo: this.repo,
      path,
      ref: sha
    })

    if (!('content' in response.data)) return ''

    const buff = Buffer.from(response.data.content, 'base64')
    return buff.toString('utf-8')
  }

  getRepoInfo(): RepoInfo {
    return getRepoInfo()
  }

  async getPrInfo(prNumber: number): Promise<PrInfo> {
    const prInfo = await this.octokit.pulls.get({
      owner: this.owner,
      repo: this.repo,
      pull_number: prNumber
    })
    return {
      baseSha: prInfo.data.base.sha,
      baseBranch: prInfo.data.base.ref
    }
  }

  async updatePrComment(commentId: number, body: string): Promise<void> {
    await this.octokit.issues.updateComment({
      owner: this.owner,
      repo: this.repo,
      comment_id: commentId,
      body
    })
  }

  async createPrComment(prNumber: number, body: string): Promise<void> {
    await this.octokit.issues.createComment({
      owner: this.owner,
      repo: this.repo,
      issue_number: prNumber,
      body
    })
  }

  async getPrBotComments(prNumber: number): Promise<PrComment[]> {
    const issueComments = await this.octokit.issues.listComments({
      owner: this.owner,
      repo: this.repo,
      issue_number: prNumber
    })

    const existingBotComments = issueComments.data.filter(
      comment => comment.user?.login === 'github-actions[bot]'
    )

    return existingBotComments.map(comment => ({
      id: comment.id,
      body: comment.body!
    }))
  }
}
