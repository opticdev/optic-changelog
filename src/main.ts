import * as core from '@actions/core'
import * as github from '@actions/github'
import {runOpticChangelog} from './changelog'
import {getJobInputs, getRepoInfo, GitHubRepository} from './github'

async function run(): Promise<void> {
  try {
    const {repoToken, subscribers, opticSpecPath} = getJobInputs()

    if (!repoToken) {
      throw new Error(
        'Please provide a GitHub token. Set one with the repo-token input or GITHUB_TOKEN env variable.'
      )
    }

    const octokit = github.getOctokit(repoToken)

    const {prNumber, owner, repo, headSha} = getRepoInfo()

    // We exit quietly when it's not a pull request
    if (!prNumber) {
      core.info('Not a pull request')
      return
    }

    const gitHubRepo = new GitHubRepository(octokit, owner, repo)
    const {baseSha, baseBranch} = await gitHubRepo.getPrInfo(prNumber)

    await runOpticChangelog({
      subscribers,
      opticSpecPath,
      gitHubRepo,
      headSha,
      baseBranch,
      baseSha,
      prNumber
    })
  } catch (error) {
    core.setFailed(error.message)
  }
}

// Don't auto-execute in the test environment
if (process.env['NODE_ENV'] !== 'test') {
  run()
}
