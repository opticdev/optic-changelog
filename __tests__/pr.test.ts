import * as fs from 'fs'
import {generateCommentBody, getMetadata, setMetadata, isOpticComment} from '../src/pr'
import {generateEndpointChanges} from '@useoptic/changelog'

const bodyWithData = `Body text here

<!-- optic = {"name":"Jane Doe"} -->`

const updatedBody = `Body text here

<!-- optic = {"name":"Jane Doe","email":"jdoe@example.com"} -->`

describe('getMetadata', () => {
  it('pulls metadata from the body text', () => {
    const result = getMetadata(bodyWithData)
    expect(result).toEqual({
      name: 'Jane Doe'
    })
  })
})

describe('setMetadata', () => {
  it('adds metadata to a body', () => {
    const data = {
      name: 'Jane Doe'
    }
    const result = setMetadata('Body text here', data)
    expect(result).toEqual(bodyWithData)
  })

  it('updates metadata in a body', () => {
    expect(
      setMetadata(bodyWithData, {
        email: 'jdoe@example.com'
      })
    ).toEqual(updatedBody)
  })
})

describe('isOpticComment', () => {
  it('recognizes Optic comment', () => {
    expect(isOpticComment(bodyWithData)).toBeTruthy
  })

  it('skips comments that are not Optic comments', () => {
    expect(isOpticComment('No Optic metadata')).toBeFalsy
  })
})

const specs = [
  {
    name: "add new endpoint",
    file: "./__tests__/specs/add-new-endpoint.json",
    sinceCreatedAt: null,
    sinceBatchCommitId: null
  },
  {
    name: "add optional response field",
    file: "./__tests__/specs/add-optional-res-field.json",
    sinceCreatedAt: "2021-04-07T15:13:51.698Z",
    sinceBatchCommitId: "9c60b8fb-faec-43f3-a0fb-171306f98d61"
  },
  {
    name: "add required response field",
    file: "./__tests__/specs/add-required-res-field.json",
    sinceCreatedAt: "2021-04-07T15:11:49.282Z",
    sinceBatchCommitId: "6c785ae0-ac1a-4c36-9fc9-e15c587459f8"
  },
  {
    name: "add response status code",
    file: "./__tests__/specs/add-res-status-code.json",
    sinceCreatedAt: "2021-04-07T15:16:53.719Z",
    sinceBatchCommitId: "9ebee98a-77a0-4dfb-b240-966eb610274b"
  },
  {
    name: "update optional response field",
    file: "./__tests__/specs/update-optional-res-field.json",
    sinceCreatedAt: "2021-04-07T15:15:16.190Z",
    sinceBatchCommitId: "5fb6357e-fc7a-4fac-8f00-db1589061e85"
  },
  {
    name: "add endpoint to existing spec",
    file: "./__tests__/specs/add-endpoint-to-existing-spec.json",
    sinceCreatedAt: "2021-04-07T15:20:11.649Z",
    sinceBatchCommitId: "c03ccae6-f3c9-4d9c-a3b6-3d710dbdb4ec"
  },
  // Use an existing spec but pick last batch commit
  {
    name: "no changes",
    file: "./__tests__/specs/add-endpoint-to-existing-spec.json",
    sinceCreatedAt: "2021-04-07T15:52:15.419Z",
    sinceBatchCommitId: "42355178-d7d7-4510-a261-bf7f579d71a3"
  },
  {
    name: "complex changes",
    file: "./__tests__/specs/complex.json",
    sinceCreatedAt: "2021-04-07T15:52:15.419Z",
    sinceBatchCommitId: "42355178-d7d7-4510-a261-bf7f579d71a3"
  },
]

const specData = specs.map(spec => {
  const currentEvents = JSON.parse(fs.readFileSync(spec.file).toString('utf-8'))
  const initialEvents = JSON.parse(fs.readFileSync(`.${spec.file.split("\.")[1]}--initial.json`).toString('utf-8'))
  return { initialEvents, currentEvents, name: spec.name }
})

specData.forEach(({initialEvents, currentEvents, name}) => {
  describe(`Comment for spec ${name}`, () => {
    let dateSpy

    beforeAll(() => {
      dateSpy = jest.spyOn(Date.prototype, 'toISOString').mockImplementation(() => "2021-04-07T19:42:18.551Z")
    })

    afterAll(() => {
      dateSpy.mockRestore()
    })

    it("renders the changelog comment", async () => {
      const changes = await generateEndpointChanges(initialEvents, currentEvents)
      const results = generateCommentBody({
        data: changes.data.endpointChanges
      }, [])
      expect(results).toMatchSnapshot()
    })
  })
})