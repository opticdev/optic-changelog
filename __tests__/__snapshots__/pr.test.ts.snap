// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment for spec add endpoint to existing spec renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 1
* Endpoints updated: 0

Last updated: 2021-04-07 19:42:18
"
`;

exports[`Comment for spec add new endpoint renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 1
* Endpoints updated: 0

Last updated: 2021-04-07 19:42:18
"
`;

exports[`Comment for spec add optional response field renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 0
* Endpoints updated: 1

Last updated: 2021-04-07 19:42:18
"
`;

exports[`Comment for spec add required response field renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 0
* Endpoints updated: 1

Last updated: 2021-04-07 19:42:18
"
`;

exports[`Comment for spec add response status code renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 0
* Endpoints updated: 1

Last updated: 2021-04-07 19:42:18
"
`;

exports[`Comment for spec complex changes renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 2
* Endpoints updated: 0

Last updated: 2021-04-07 19:42:18
"
`;

exports[`Comment for spec no changes renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 1
* Endpoints updated: 0

Last updated: 2021-04-07 19:42:18
"
`;

exports[`Comment for spec update optional response field renders the changelog comment 1`] = `
"## Optic Changelog
  
* Endpoints added: 0
* Endpoints updated: 1

Last updated: 2021-04-07 19:42:18
"
`;
